src/main/java/com/pinterest/secor/reader/SecorKafkaMessageIterator.java

```diff
@@ -0,0 +1,145 @@
+        optionalConfig(config.getSslTruststorePassword(), conf -> props.put("ssl.truststore.password", conf));
+        optionalConfig(config.getIsolationLevel(), conf -> props.put("isolation.level", conf));
+        optionalConfig(config.getMaxPollRecords(), conf -> props.put("max.poll.records", conf));
+        optionalConfig(config.getSaslClientCallbackHandlerClass(), conf -> props.put("sasl.client.callback.handler.class", conf));
+        optionalConfig(config.getSaslJaasConfig(), conf -> props.put("sasl.jaas.config", conf));
+        optionalConfig(config.getSaslKerberosServiceName(), conf -> props.put("sasl.kerberos.service.name", conf));
+        optionalConfig(config.getSaslLoginCallbackHandlerClass(), conf -> props.put("sasl.login.callback.handler.class", conf));
+        optionalConfig(config.getSaslLoginClass(), conf -> props.put("sasl.login.class", conf));
+        optionalConfig(config.getSaslMechanism(), conf -> props.put("sasl.mechanism", conf));
+        optionalConfig(config.getSecurityProtocol(), conf -> props.put("security.protocol", conf));
+        optionalConfig(config.getSslEnabledProtocol(), conf -> props.put("ssl.enabled.protocols", conf));
+        optionalConfig(config.getSslKeystoreType(), conf -> props.put("ssl.keystore.type", conf));
+        optionalConfig(config.getSslProtocol(), conf -> props.put("ssl.protocol", conf));
+        optionalConfig(config.getSslProvider(), conf -> props.put("ssl.provider", conf));
+        optionalConfig(config.getSslTruststoreType(), conf -> props.put("ssl.truststore.type", conf));
+
+        mZookeeperConnector = new ZookeeperConnector(config);
+        mRecordsBatch = new ArrayDeque<>();
+        mKafkaConsumer = new KafkaConsumer<>(props);
+        mKafkaConsumer.subscribe(Pattern.compile(config.getKafkaTopicFilter()), new ConsumerRebalanceListener() {
+            @Override
+            public void onPartitionsRevoked(Collection<TopicPartition> collection) {
+                // TODO: Validate if secor needs something done here
+                for (TopicPartition topicPartition : collection) {
+                    LOG.debug("Topic partition {} revoked from consumer", topicPartition);
+                }
+            }
+
+            @Override
+            public void onPartitionsAssigned(Collection<TopicPartition> collection) {
+                Map<TopicPartition, Long> committedOffsets = getCommittedOffsets(collection);
+                committedOffsets.forEach(((topicPartition, offset) -> {
+                    if (offset == -1) {
+                        if (offsetResetConfig.equals("earliest")) {
+                            mKafkaConsumer.seekToBeginning(collection);
+                        } else if (offsetResetConfig.equals("latest")) {
+                            mKafkaConsumer.seekToEnd(collection);
+                        }
+                    } else {
+                        LOG.debug("Seeking {} to offset {}", topicPartition, Math.max(0, offset));
```

Comment:
Is this a kafka bug?  I would imagine the kafka consumer will reset to the correct offset after partition reassignment.  And also do we need to clear any data structure and local data file because of the offset change?

Reply:
We need this because new kafka consumers don't support storing offsets in zookeeper. After an assignment, we need to manually seek to the offsets stored there[0]. The decision to keep using zookeeper came from two major points:
- People will be able to upgrade to a new secor version without having to migrate their commited offsets to kafka storage
- New kafka consumers commit model doesn't fit well with secor, in a nutshell, `consumer.commit` can throw a non retryable exception called `CommitFailedException`[1]. If this happens, we can either ignore it and end up with duplicate data(partitions were assigned to another consumer) or we can try to undo our action(deleting the file from s3). I thought that implementing this functionality is risky and not worth it since we can just keep using the "old" model of commiting to zookeeper.

[0] - https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.html
[1] - https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/consumer/CommitFailedException.html

Reply:
Partition rebalance also happen in the old days for the old consumers, in that case, we just let it continue.  In the end, only one worker is able to upload/persist the datafile for a given partition because each upload has to make sure the pre-condition satisfied (starting offset on S3 for the given partition matches with its own data structure)

In the new consumer case, do you store the offsets in dual storage (both ZK and kafka topic)?  What's the value of offsets.storage and dual.commit.enabled parameters in your case?  I would think kafka would still maintain the offsets in __consumer_topics during rebalance (eliminate the need to manually call seek) if either offsets.storage was set to kafka or dual committing is turned on.

Reply:
Currently, I'm ignoring both `offsets.storage` and `dual.commit.enabled` but I'll fix it. If `offset.storage` is set to kafka I won't manually seek the offsets. I just need to add a config check so that the user does not set `offset.storage` to kafka and `dual.commit.enabled` to false.

I thought that I wasn't going to be able to commit offsets to kafka, but I was wrong. Secor always commits to zookeeper and this prevents the clients from uploading duplicate offsets. Even if zookeeper commit fails, the second upload should overwrite the first because the filename will be same. Is this correct or have I missed something?

Reply:
Or maybe I'm just confused about the overall model here.  If I'm currently using the legacy iterator (or well, an older version Secor) and zookeeper offset storage, what's the right way to upgrade to the using the newer iterator? Should I do a two-step upgrade where I first upgrade to "legacy iterator, kafka storage, dual commit", then to "secor iterator, kafka storage, dual commit"?  Or a three-step upgrade going to "legacy, kafka, dual" -> "legacy, kafka, no dual" -> "secor, kafka, no dual"?  Or just a one-step upgrade?